pkg/controller/replicaset/replica_set.go

```diff
@@ -248,7 +271,36 @@ func (rsc *ReplicaSetController) updateRS(old, cur interface{}) {
 	if *(oldRS.Spec.Replicas) != *(curRS.Spec.Replicas) {
 		klog.V(4).Infof("%v %v updated. Desired pod count change: %d->%d", rsc.Kind, curRS.Name, *(oldRS.Spec.Replicas), *(curRS.Spec.Replicas))
 	}
-	rsc.enqueueReplicaSet(cur)
+	rsc.enqueueRS(curRS)
+}
+
+func (rsc *ReplicaSetController) deleteRS(obj interface{}) {
+	rs, ok := obj.(*apps.ReplicaSet)
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("couldn't get object from tombstone %#v", obj))
+			return
+		}
+		rs, ok = tombstone.Obj.(*apps.ReplicaSet)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("tombstone contained object that is not a ReplicaSet %#v", obj))
+			return
+		}
+	}
+
+	key, err := controller.KeyFunc(rs)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %#v: %v", rs, err))
+		return
+	}
+
+	klog.V(4).Infof("Deleting %s %q", rsc.Kind, key)
+
+	// Delete expectations for the ReplicaSet so if we create a new one with the same name it starts clean
+	rsc.expectations.DeleteExpectations(key)
```

Comment:
also need to consider the case in `updateRS` where `oldRS.UID != curRS.UID`. You're not guaranteed to see a deleteRS call for every delete/recreate case.

Reply:
not even tombstone? I thought `DeletionHandlingMetaNamespaceKeyFunc` and the internal queue would generate artificial delete event in case it missed the real delete event and already saw new RS with new UID

Reply:
no, a tombstone is sent if a relist is missing the object entirely. if the relist has the object with the same name with a new uid, an update is sent

Reply:
is that intentional?
https://github.com/kubernetes/kubernetes/blob/9e4be54/staging/src/k8s.io/client-go/tools/cache/delta_fifo.go#L628-L632

states

```
// DeletedFinalStateUnknown is placed into a DeltaFIFO in the case where
// an object was deleted but the watch deletion event was missed. In this
// case we don't know the final "resting" state of the object, so there's
// a chance the included `Obj` is stale.
type DeletedFinalStateUnknown struct {
```

yet I see the deltaFIFo backing what you say in

Reply:
deltafifo doesn't understand uids, and the cache key doesn't include the uid. a delete/recreate is not the same as a delete/final-state-unknown from deltafifo's perspective

Reply:
I guess deltafifo needs to have the same key anyways or it could reorder the events. I meant to fix the processor - let me push the change to be more concrete